[Attention] Flash Attention 3 - fp8

mickaelseznec opened the pull request:
This PR add support for FP8 KV cache with FlashAttention3 (related PR in flash-attn [here](https://github.com/vllm-project/flash-attention/pull/50)) cc @LucasWilkinson Please do not merge this PR as long as it's not referencing vllm-project/flash-attention yet.

FlashAttention (contrary to FlashInfer) does attention with all Q, K and V in FP8.
The performance is usually better than FlashInfer FP8 KV and FlashAttention 3 with bf16.

I added support for v0 and v1 + some unit testing.

Note that I've added a trick for checkpoints not providing q_scale and reuse the k_scale (with is something TRTLLM does fwiw).

Also: I added a small QoS improvement when debugging v1: workers send back their traceback when they raise an exception.

mickaelseznec committed:
[Feature] Add support for Flash-Attn FP8

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

mickaelseznec committed:
[Feature] executors send back traceback

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

mickaelseznec committed:
[FIXME] temp. use up-to-date flashattn

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

mickaelseznec committed:
feat: add unit test for FA3 FP8

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

mickaelseznec committed:
fix: ruff

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

mickaelseznec commented:
CI failing because` vllm/tests/entrypoints/openai/test_accuracy.py` from [here](https://github.com/vllm-project/vllm/blob/main/.buildkite/run-tpu-test.sh#L22) doesn't exist.

@robertgshaw2-redhat any idea how should I fix? Just rename in run-tpu-test.sh? (@NickLucche you moved the file)

NickLucche commented:
This is a known issue, PR addressing it here https://github.com/vllm-project/vllm/pull/13898. It won't block your PR.

NickLucche commented:
I see there's some other problem with building the image, but likely CI just needs another spin

LucasWilkinson commented:
@mickaelseznec apologies for the delay, https://github.com/vllm-project/flash-attention/pull/50#pullrequestreview-2675615072 has been merged, you can now point to vllm_flash_attn

We will need to populate the sccache on the server to get it through the CI, I can help with this once the tag is updated :+1:

LucasWilkinson commented on tests/kernels/test_flash_attn.py:
```diff
@@ -223,10 +261,30 @@ def test_varlen_with_paged_kv(
+
+        q_descale = q_scale.expand((num_seqs, num_kv_heads))
+        k_descale = k_scale.expand((num_seqs, num_kv_heads))
+        v_descale = v_scale.expand((num_seqs, num_kv_heads))
```
nit: could we maybe test per-head scales here too?, i.e. also test with non-zero strides

LucasWilkinson commented on vllm/platforms/cuda.py:
```diff
@@ -240,15 +240,6 @@ def get_attn_backend_cls(cls, selected_backend, head_size, dtype,
                 "Cannot use FlashAttention-2 backend for dtype other than "
                 "torch.float16 or torch.bfloat16.")
             target_backend = _Backend.XFORMERS
-        elif kv_cache_dtype is not None and \
```
we should keep this check but restrict it to FA2, i.e. check `get_flash_attn_version() != 2` (`get_flash_attn_version()` is in `vllm/attention/backends/utils.py`)

mickaelseznec replied:
Agree that this might be improved, but I can't directly import `get_flash_attn_version` because of circular dependency.

Do you prefer if I move that function in another file? `vllm/attention/backends/versions.py` for example?

LucasWilkinson replied:
> Do you prefer if I move that function in another file? vllm/attention/backends/versions.py for example?

sure, maybe move it to:
```
vllm/attention/utils/fa_support.py
```
for now, since there is `is_flash_attn_mla_supported()` function that may come in https://github.com/vllm-project/vllm/pull/14258 so this could be a spot for both of those

mickaelseznec replied:
I had to move it to `vllm/fa_utils.py` because of how `vllm/attention/__init__.py` imports a bunch of stuff for convenience.

LucasWilkinson reviewed (commented):
Thanks for the contribution! Looks clean 😄, ill approve once we can get it updated to use `vllm_flash_attn`, added a couple comments

mickaelseznec committed:
feat: update flash-attention reference

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

mickaelseznec committed:
Merge branch 'main' into mseznec/flash-attention-fp8

mickaelseznec committed:
refactor: get_flash_attn in dedicated file

this is needed to avoid circular dependencies now that we want to get
the flash_attn_version directly in platforms/cuda.py to check if fp8
flash_attn is actually available

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

mickaelseznec committed:
Merge branch 'main' into mseznec/flash-attention-fp8

mickaelseznec committed:
refactor: get_flash_attn_version

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

LucasWilkinson commented:
apologies for the delay, the CI should be working now. There appears to be failing kernel tests

mickaelseznec committed:
fix: tests

increase flash_attention unit test tolerance

Signed-off-by: Mickael Seznec <mickael@mistral.ai>

JaheimLee commented:
Hi, it seems there is no new nightly wheel since this PR. Is there anything wrong? @LucasWilkinson
